lnwallet/chanfunding/psbt_assembler.go

```diff
@@ -0,0 +1,135 @@
+package chanfunding
+
+import (
+	"fmt"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/btcsuite/btcutil/psbt"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/keychain"
+)
+
+// PsbtIntent is an intent created by the PsbtIntent which represents a
+// funding output to be created by a PSBT. This might be used when a hardware
+// wallet, or a channel factory is the entity crafting the funding transaction,
+// and not lnd.
+type PsbtIntent struct {
+	ShimIntent
+
+	// PeerKey is the peer the channel opening negotiation has been started
+	// with.
+	PeerKey *btcec.PublicKey
+
+	// Psbt is the parsed version of the current PSBT. This can be in two
+	// stages: If the user has not yet provided any PSBT, this is nil. Once
+	// the user sends us a first funded PSBT, we verify that we have a valid
+	// transaction that sends to the channel output PK script and has an
+	// input large enough to pay for it. We keep this verified but not yet
+	// signed version around until the fully signed transaction is submitted
+	// by the user. At that point we make sure the inputs and outputs
+	// haven't changed to what was previously verified. Only witness data
+	// should be added after the verification process.
+	Psbt *psbt.Packet
+}
+
+func (i *PsbtIntent) CompileFundingTx() (*wire.MsgTx, error) {
+	if i.Psbt == nil {
+		return nil, fmt.Errorf("PSBT not ready")
+	}
+
+	// Make sure the PSBT can be finalized and is complete.
+	err := psbt.MaybeFinalizeAll(i.Psbt)
+	if err != nil {
+		return nil, err
+	}
+	if !i.Psbt.IsComplete() {
+		return nil, fmt.Errorf("PSBT not complete")
+	}
+
+	// Extract the final transaction.
+	fundingTx, err := psbt.Extract(i.Psbt)
+	if err != nil {
+		return nil, fmt.Errorf("unable to extract funding TX: %v", err)
+	}
+
+	// Identify our funding outpoint now that we know everything's ready.
+	_, txOut, err := i.FundingOutput()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get funding output: %v", err)
+	}
+	ok, idx := input.FindScriptOutputIndex(fundingTx, txOut.PkScript)
+	if !ok {
+		return nil, fmt.Errorf("funding output not found in PSBT")
+	}
+	i.chanPoint = &wire.OutPoint{
+		Hash:  fundingTx.TxHash(),
+		Index: idx,
+	}
+
+	return fundingTx, nil
+}
+
+func (i *PsbtIntent) BindKeys(localKey *keychain.KeyDescriptor,
+	remoteKey *btcec.PublicKey) {
+
+	i.localKey = localKey
```

Comment:
Is this needed given that we wrap the shim intent?

Reply:
Yes, the `BindKeys` only also exists on the `FullIntent`, not the `ShimIntent`.